### ISSUE TITLE
version 2.4.7 many threads waiting on getConnectiom method

### ISSUE THREAD

**Issue by wplatform** [brettwooldridge/HikariCP#714]
"ddal-query-executor-535" - Thread t@625
   java.lang.Thread.State: TIMED_WAITING
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for <5612c816> (a com.zaxxer.hikari.util.QueuedSequenceSynchronizer$Synchronizer)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedLongSynchronizer.doAcquireSharedNanos(AbstractQueuedLongSynchronizer.java:815)
	at java.util.concurrent.locks.AbstractQueuedLongSynchronizer.tryAcquireSharedNanos(AbstractQueuedLongSynchronizer.java:1106)
	at com.zaxxer.hikari.util.QueuedSequenceSynchronizer.waitUntilSequenceExceeded(QueuedSequenceSynchronizer.java:92)
	at com.zaxxer.hikari.util.ConcurrentBag.borrow(ConcurrentBag.java:169)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:165)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:147)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:99)
	at com.openddal.repo.JdbcRepository$1.getConnection(JdbcRepository.java:330)
	at com.openddal.repo.tx.ConnectionHolder.getRawConnection(ConnectionHolder.java:189)
	at com.openddal.repo.tx.ConnectionHolder.getConnection(ConnectionHolder.java:75)
	at com.openddal.repo.JdbcWorker.doGetConnection(JdbcWorker.java:74)
	at com.openddal.repo.JdbcQueryWorker.executeQuery(JdbcQueryWorker.java:51)
	at com.openddal.repo.JdbcQueryWorker.call(JdbcQueryWorker.java:40)
	at com.openddal.repo.JdbcQueryWorker.call(JdbcQueryWorker.java:32)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

Caused by: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 30000ms.
	at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:555) ~[HikariCP-2.4.7.jar:na]
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:188) ~[HikariCP-2.4.7.jar:na]
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:147) ~[HikariCP-2.4.7.jar:na]
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:99) ~[HikariCP-2.4.7.jar:na]
	at com.openddal.repo.JdbcRepository$1.getConnection(JdbcRepository.java:330) ~[openddal-engine-1.2.1-SNAPSHOT.jar:na]
	at com.openddal.repo.tx.ConnectionHolder.getRawConnection(ConnectionHolder.java:189) ~[openddal-engine-1.2.1-SNAPSHOT.jar:na]
	at com.openddal.repo.tx.ConnectionHolder.getConnection(ConnectionHolder.java:75) ~[openddal-engine-1.2.1-SNAPSHOT.jar:na]
	at com.openddal.repo.JdbcWorker.doGetConnection(JdbcWorker.java:74) ~[openddal-engine-1.2.1-SNAPSHOT.jar:na]
	at com.openddal.repo.JdbcQueryWorker.executeQuery(JdbcQueryWorker.java:51) ~[openddal-engine-1.2.1-SNAPSHOT.jar:na]
	at com.openddal.repo.JdbcQueryWorker.call(JdbcQueryWorker.java:40) ~[openddal-engine-1.2.1-SNAPSHOT.jar:na]
	at com.openddal.repo.JdbcQueryWorker.call(JdbcQueryWorker.java:32) ~[openddal-engine-1.2.1-SNAPSHOT.jar:na]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_51]
	... 3 common frames omitted

**Comment by wplatform** [comment]
if  put synchronized on method, the problem will not appear.

public synchronized Connection getConnection(Options options) {
                DataSource ds = getDataSourceByShardName(options.shardName);
                try {
                    return ds.getConnection();
                } catch (SQLException e) {
                    throw DbException.convert(e);
                }
            }

**Comment by brettwooldridge** [comment]
What "problem"?  In one case the threads are waiting on the QueuedSynchronizer and in the other (synchronized) case they will block on the method lock.

**Comment by wplatform** [comment]
Multi-threaded concurrently get  connection from the database, found a large number of threads waiting on QueuedSynchronizer, if put synchronized on a method, there is no problem, it’s the getConnection method is thread-safe?

**Comment by brettwooldridge** [comment]
What "problem"?  In one case the threads are waiting on the QueuedSynchronizer and in the other (synchronized) case they will block on the method lock.

**Comment by brettwooldridge** [comment]
It's perfectly normal.  HikariCP uses the synchronizer which is a special kind of lock.  If the number of threads outnumber the number of connections, you'll find them parked on this lock.  When a connection becomes available a waiting thread will become unpacked and acquire the connection.

**Comment by wplatform** [comment]
I don't think so. It might be a bug, in my case, get connected and close the connection is not the same thread, lead to the connection object cannot return to local thread pool.And Shared pool without connection object, connection object cannot be closed.

**Comment by brettwooldridge** [comment]
@wplatform I have [uploaded a snapshot build](https://oss.sonatype.org/content/repositories/snapshots/com/zaxxer/HikariCP/2.5.1-SNAPSHOT/HikariCP-2.5.1-20160913.172644-1.jar), can you give it a try?  It is Java 8 only.  If you can test under Java 8, at least to verify the fix, I will also propagate the change back to the Java 7 branch.

**Comment by johnou** [comment]
@wplatform any luck with testing the snapshot?
